pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,311 @@
+	}
+
+	err = ensureDisk(ctx, gcpClient, opt, true)
+	if err != nil {
+		return err
+	}
+
+	instance, err := ensureComputeInstance(ctx, logger, bastion, gcpClient, opt, true)
+	if err != nil {
+		return err
+	}
+
+	// check if the instance already exists and has an IP
+	endpoints, err := getInstanceEndpoints(instance)
+	if err != nil {
+		return err
+	}
+
+	if !endpoints.Ready() {
+		return &ctrlerror.RequeueAfterError{
+			// requeue rather soon, so that the user (most likely gardenctl eventually)
+			// doesn't have to wait too long for the public endpoint to become available
+			RequeueAfter: 5 * time.Second,
+			Cause:        fmt.Errorf("bastion instance has no public/private endpoints yet"),
+		}
+	}
+
+	err = controller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.Client(), bastion, func() error {
+		bytes, _ := marshalProviderStatus(opt.Zone)
+		if err != nil {
+			return err
+		}
+		bastion.Status.ProviderStatus = &runtime.RawExtension{Raw: bytes}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to store status.providerStatus for zone: %s", opt.Zone)
+	}
+
+	// once a public endpoint is available, publish the endpoint on the
+	// Bastion resource to notify upstream about the ready instance
+	return controller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.Client(), bastion, func() error {
+		bastion.Status.Ingress = *endpoints.public
+		return nil
+	})
+
+}
+
+func ensureFirewallRules(ctx context.Context, gcpclient gcpclient.Interface, opt *Options) error {
+	firewallList := []*compute.Firewall{IngressAllowSSH(opt), EgressDenyAll(opt), EgressAllowOnly(opt)}
+
+	for _, item := range firewallList {
+		if err := createFirewallRuleIfNotExist(ctx, gcpclient, opt, item); err != nil {
+			return err
+		}
+	}
+
+	firewall, err := getFirewallRule(ctx, gcpclient, opt, IngressAllowSSH(opt).Name)
+	if err != nil {
+		return fmt.Errorf("%w, could not get firewall rule", err)
+	}
+
+	currentCIDRs := firewall.SourceRanges
+	wantedCIDRs := opt.CIDRs
+
+	if !reflect.DeepEqual(currentCIDRs, wantedCIDRs) {
+		return patchFirewallRule(ctx, gcpclient, opt, IngressAllowSSH(opt).Name)
+	}
+
+	return nil
+}
+
+func ensureComputeInstance(ctx context.Context, logger logr.Logger, bastion *extensionsv1alpha1.Bastion, gcpclient gcpclient.Interface, opt *Options, allowRecursion bool) (*compute.Instance, error) {
+	instance, err := getBastionInstance(ctx, gcpclient, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	if instance != nil {
+		return instance, nil
+	}
+
+	logger.Info("running new bastion compute instance instance")
+	computeInstance := computeInstanceDefine(opt, bastion.Spec.UserData)
+	_, err = gcpclient.Instances().Insert(opt.ProjectID, opt.Zone, computeInstance).Context(ctx).Do()
+	if err != nil {
+		return nil, fmt.Errorf("%w, failed to create bastion compute instance", err)
+	}
+
+	if allowRecursion {
+		return ensureComputeInstance(ctx, logger, bastion, gcpclient, opt, false)
+	}
```

Comment:
I had checked with Andrei as he was sick last week, the recursion here is used to prevent infinitive loop.

Reply:
@andrei-panov my question was rather: why is there a need to call this function recursively? Isn't it sufficient to just run
```go
	return getBastionInstance(ctx, gcpclient, opt)
```
after the insert?

Reply:
Hi @petersutter, I removed the recursion and add a call for getBastionInstance for the second time. `return getBastion...` is not sufficient because we want to throw the error if the instance is still not there. Same for `getDisk`

Reply:
yes right, that's fine